src/segments/umbraco.go

```diff
@@ -0,0 +1,231 @@
+	env   platform.Environment
+
+	FoundUmbraco    bool
+	IsModernUmbraco bool
+	IsLegacyUmbraco bool
+	Version         string
+}
+
+const (
+	umbracoFolderName = "umbraco"
+	umbracoWebConfig  = "web.config"
+)
+
+type CSProj struct {
+	PackageReferences []struct {
+		Name    string `xml:"Include,attr"`
+		Version string `xml:"Version,attr"`
+	} `xml:"ItemGroup>PackageReference"`
+}
+
+type WebConfig struct {
+	AppSettings []struct {
+		Key   string `xml:"key,attr"` // TODO: What happens if the web.config has the attribute as uppercase Key="" ?
+		Value string `xml:"value,attr"`
+	} `xml:"appSettings>add"`
+}
+
+type FindUmbracoResult struct {
+	FoundUmbracoFolder bool
+	FoundWebConfig     bool
+	FoundCSProj        bool
+	FilePath           string
+}
+
+func (u *Umbraco) Enabled() bool {
+	u.env.Debug("UMBRACO: Checking if we enable segment")
+
+	// Checks if the current folder contains an umbraco folder
+	// If not it then checks each parent until the root
+	findUmbracoResults, err := u.TryFindUmbracoInParentDirsOrSelf()
+
+	if err != nil {
+		u.env.Debug("UMBRACO: Error while searching for Umbraco folder and files")
+		u.env.Debug(err.Error())
+		return false
+	}
+
+	if !findUmbracoResults.FoundUmbracoFolder {
+		u.env.Debug("UMBRACO: No Umbraco folder found")
+		u.FoundUmbraco = false
+		return false
+	}
+
+	// We have found an Umbraco folder between CWD and root
+	// Along with finding a web.config file at the same level as the Umbraco folder
+	// OR one or more .csproj files at the same level as the Umbraco folder
+
+	// Modern .NET Core based Umbraco
+	if findUmbracoResults.FoundCSProj {
+		u.env.Debug("UMBRACO: Checking for modern Umbraco as we have found .csproj file")
+		return u.TryFindModernUmbraco(findUmbracoResults.FilePath)
+	}
+
+	// Legacy .NET Framework based Umbraco
+	if findUmbracoResults.FoundWebConfig {
+		u.env.Debug("UMBRACO: Checking for legacy Umbraco as we have found web.config file")
+		return u.TryFindLegacyUmbraco(findUmbracoResults.FilePath)
+	}
+
+	// If we have got here then neither modern or legacy Umbraco was NOT found
+	u.FoundUmbraco = false
+	return false
+}
+
+func (u *Umbraco) Template() string {
+	return "{{.Version}} "
+}
+
+func (u *Umbraco) Init(props properties.Properties, env platform.Environment) {
+	u.props = props
+	u.env = env
+}
+
+func (u *Umbraco) TryFindUmbracoInParentDirsOrSelf() (*FindUmbracoResult, error) {
+	defer u.env.Trace(time.Now(), "UMBRACO: Checking for an umbraco folder & files in current or any parent folders until root")
+	currentFolder := u.env.Pwd()
+	results := FindUmbracoResult{}
+
+	for {
+		// Check if a directory named "Umbraco" exists in the current directory
+		files := u.env.LsDir(currentFolder)
+
+		// Have to loop over each item found in the current folder
+		for _, file := range files {
+			// Check if the item is a folder AND it matches 'Umbraco' regardless of casing
+			if file.IsDir() && strings.EqualFold(file.Name(), umbracoFolderName) {
```

Comment:
if you invert this one you can break when there's no folder found instead of having to carry the check down all the way to line 131.

Reply:
Not sure if this would be best to do. As looking at `env.LsDir()` which uses `os.ReadDir()` and as this seems to sort all results regardless if its a folder or a file alphabetically so I can't assume if its not the Umbraco folder to break out early as it could be the web.config or a *.csroj file

I can break each inside if though, which I'll rectify

Reply:
I meant continue 😓

Reply:
Want me to add in `continue` into each of these cases?

Reply:
I had a go at it and I have this:

```go
func (u *Umbraco) Enabled() bool {
	var location string

	folders := []string{"umbraco", "Umbraco"}
	for _, folder := range folders {
		if file, err := u.env.HasParentFilePath(folder); err == nil {
			location = file.ParentFolder
			break
		}
	}

	if len(location) == 0 {
		u.env.Debug("No umbraco folder found in parent directories")
		return false
	}

	files := u.env.LsDir(location)

	for _, file := range files {
		if file.IsDir() {
			continue
		}

		if strings.EqualFold(file.Name(), "web.config") {
			return u.TryFindLegacyUmbraco(filepath.Join(location, file.Name()))
		}

		if strings.EqualFold(filepath.Ext(file.Name()), ".csproj") {
			return u.TryFindModernUmbraco(filepath.Join(location, file.Name()))
		}
	}

	return false
}
```

But that would imply that we always look for a config file at the umbraco folder's location.

Reply:
LGTM and thanks for simplifying this @JanDeDobbeleer 

Yes its the assumption that does need to be made. Just finding an Umbraco folder on it's own does not indicate that it is an Umbraco project it has to be the existence of the *.csproj with the nuget package installed or the web.config with the Umbraco version in the AppSetting Key.

So I'd say lets go for that
![LGTM](https://media.giphy.com/media/NytMLKyiaIh6VH9SPm/giphy.gif)

Reply:
I'll get that done today! After dinner.

Reply:
Let me know if you want me to implement what you suggested, if you are too busy @JanDeDobbeleer